kubectl create: enforce namespace during dry run

busser opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

When creating a role with `kubectl create role`, any specified namespace is ignored if the command is a dry run. This is unexpected behavior.

For example, the following command yields the following output:

```bash
kubectl create role foo --verb=get --resource=pod --namespace=bar --output=yaml --dry-run
```

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  creationTimestamp: null
  name: foo
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
```

This pull-request fixes this issue. The command above now provides the expected output:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  creationTimestamp: null
  name: foo
  namespace: bar
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
```

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubectl/issues/705

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The output of a dry run of creating a resource with kubectl now always takes into account the namespace specified in the command.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

busser committed:
kubectl create role: add namespace during dry run

soltysh commented:
/ok-to-test

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_role.go:
```diff
@@ -329,6 +330,9 @@ func (o *CreateRoleOptions) RunCreateRole() error {
 	}
 	role.Name = o.Name
+	if o.NamespaceOverwritten {
+		role.Namespace = o.Namespace
```
I don't even think this condition is necessary, we should just ensure the `role.Namespace` is set.

busser replied:
That is what I did originally, but it resulted in `metadata.namespace` being explicitly set to `default` in the YAML output when the user did not use the `--namespace` option. For example:

```bash
kubectl create role foo --verb=get --resource=pod --output=yaml --dry-run
```

would now output this:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  creationTimestamp: null
  name: foo
  namespace: default
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
```

while it used to output this:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  creationTimestamp: null
  name: foo
rules:
- apiGroups:
  - ""
  resources:
  - pods
  verbs:
  - get
```

soltysh replied:
Fair, please use `EnforceNamespace` as the var name, in that case.

busser replied:
Done.

soltysh reviewed (changes requested):
Although https://github.com/kubernetes/kubectl/issues/705 uses `create role` as an example this problem applies to majority (if not all) of create commands. Can you please check all of them? Also don't forget to add tests ensuring that when run with `--dry-run` the namespace is properly set.

busser committed:
Rename NamespaceOverwritten field to EnforceNamespace

busser commented:
/hold

Putting this PR on hold while I fix the other `kubectl create` commands and add tests.

busser committed:
kubectl/cmd/create: add TestRunCreateRole unit tests

soltysh commented:
/priority backlog

busser committed:
Enforce namespace after WithNamespace is called.

When the TestConfigFlags.WithNamespace method is called, the
ClientConfig is wrapped inside a namespacedClientConfig.
Previously, calling the namespaceClientConfig.Namespace method would
return that the namespace should not be enforced. This made testing
namespace enforcement complicated. From now on, the method will
return that the namespace should be enforced.

busser committed:
Fix unit test

A change in the cmdtesting.TestFactory broke the TestSetLocalNamespace
unit test. Calling the TestFactory.WithNamespace method now enforces
the "test" namespace, which conflicts with how this test works.

This unit tests works by loading a resource from a file with its
namespace set to "existing-ns", which was overwritten by "test".

This fix corrects this issue.

busser committed:
kubectl create role: add namespace to test

The create.RunCreateRole function now enforces namespace during dry runs.
The TestCreateRole unit test has been updated to reflect that.

busser committed:
kubectl create clusterrole: refactor unit test

TestCreateClusterRole now runs as multiple subtests and uses the diff package
to print differences between the expected and actual cluster roles.

busser committed:
kubectl create cronjob: add namespace during dry run

If the namespace is to be enforced, make sure to always enforce it, even
during dry runs.

busser committed:
kubectl create cronjob: refactor unit test

TestCreateCronJob now uses the diff package to print differences
between the expected and actual cron jobs.

busser committed:
kubectl create role: rename unit tests

The tests were renamed because there could not be a TestComplete
function for both roles and cron jobs, for example. Same for
TestValidate.

busser committed:
kubectl create cronjob: add unit tests

busser committed:
kubectl create job: add namespace during dry run

If the namespace is to be enforced, make sure to always enforce it, even
during dry runs.

busser committed:
kubectl create cronjob: remove unused TestFactory

busser committed:
kubectl create cronjob: fix unit test error message

busser committed:
kubectl create job: add unit test

busser committed:
kubectl create job: remove unit test

The TestCreateRoleValidation test was replaced by TestValidateCreateRole
to match how other `kubectl create` commands are tested.

busser commented:
/test pull-kubernetes-node-e2e-containerd

busser commented:
/test pull-kubernetes-node-e2e-containerd

busser commented:
The `pull-kubernetes-node-e2e-containerd` test keeps failing. I am getting the same error messages as in issue #85040.

busser commented:
/unhold

All `kubectl create` sub-commands have been checked and, when needed, fixed.

busser commented:
/test pull-kubernetes-node-e2e-containerd

soltysh commented on staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags_fake.go:
```diff
@@ -96,7 +96,7 @@ type namespacedClientConfig struct {
 
 func (c *namespacedClientConfig) Namespace() (string, bool, error) {
-	return c.namespace, false, nil
+	return c.namespace, true, nil
```
I thought we agreed this to be:
```go
return c.namespace, len(c.namespace) > 0, nil
```
didn't we?

busser replied:
You're right.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_cronjob.go:
```diff
@@ -149,18 +150,28 @@ func (o *CreateCronJobOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, a
 	cronjob = o.createCronJob()
+	if o.EnforceNamespace {
+		cronjob.Namespace = o.Namespace
+		cronjob.Spec.JobTemplate.Namespace = o.Namespace
```
Not needed, namespace for the created job will be always inherited from the cronjob itself.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_cronjob.go:
```diff
@@ -149,18 +150,28 @@ func (o *CreateCronJobOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, a
 }
 
 func (o *CreateCronJobOptions) Validate() error {
+	if o.Name == "" {
```
`len(o.Name) == 0` is preferred, see below.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_cronjob_test.go:
```diff
@@ -112,9 +118,279 @@ func TestCreateCronJob(t *testing.T) {
+	}
+}
+
+func TestRunCreateCronJob(t *testing.T) {
```
This is mostly duplicating `TestCreateCronJob`, if you look at what `Run` in `create_cronjob.go` does it's only creating the CJ itself and I don't think we need that much of testing of the printing and all the rest of the machinery. This test has more setup code than actual code, and one of the reasons we've went with `*Options` struct was to get rid of those additional setup steps. The high level testing should happen in test-cmd, instead.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_cronjob_test.go:
```diff
@@ -112,9 +118,279 @@ func TestCreateCronJob(t *testing.T) {
+	}
+}
+
+func TestCompleteCreateCronJob(t *testing.T) {
```
The same argument as below, I don't think we need to test every single line of create commands, `Complete` method is one of those which can be easily tested in test-cmd.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_clusterrole_test.go:
```diff
@@ -149,26 +150,28 @@ func TestCreateClusterRole(t *testing.T) {
+			if test.resourceNames != "" {
+				cmd.Flags().Set("resource-name", test.resourceNames)
+			}
+			cmd.Run(cmd, []string{clusterRoleName})
```
I'd prefer we move with the testing towards the pattern from create cronjob, where you directly inject values into `*Options` struct and focus testing on the "raw meat", which in here is `RunCreateRole`.

soltysh replied:
The less of the test factory magic, the easier this testing is for newcomers to understand it.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/create/create_role_test.go:
```diff
@@ -22,7 +22,7 @@ import (
 	rbac "k8s.io/api/rbac/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
`metav1` is the proffered alias for this, since you've touched this.

soltysh reviewed (changes requested):
Also please squash your changes into a single commit.

zhouya0 commented:
@busser Seems no updates from this PR? Could I take this bug and send my PR?

busser commented:
> @busser Seems no updates from this PR? Could I take this bug and send my PR?

Sorry, I haven't taken the time to finish this. Feel free to submit your solution :)

zhouya0 commented:
> > @busser Seems no updates from this PR? Could I take this bug and send my PR?
> 
> Sorry, I haven't taken the time to finish this. Feel free to submit your solution :)

Yes, it's already been fixed and we can close this :).
/close
